### ISSUE TITLE
Add zipAllWith to Chunk

### ISSUE THREAD

**Issue by jdegoes** [zio/zio#1164]
Currently, `Chunk#zipAll` can be used to zip chunks together with a specified combiner function. However, the resulting chunk will always have a length equal to the minimum of the length of the two chunks, resulting in the "loss" of the elements in the longer chunk.

For more flexible behavior, we can add a new operator that allows zipping two chunks and dealing with any "extra" elements on the longer side.

```scala
trait Chunk[A] {
  /**
   * Zips all elements into a Chunk that has the maximum length of this chunk
   * and the specified chunk.
  def zipAllWith[@specialized B, @specialized C](that: Chunk[B])(left: A => C, right: B => C)(both: (A, B) => C)): Chunk[C]
}
```

Where both sides have an element, `both` is used to synthesize elements. Where the left side is shorter than the right side, then `right` is used to synthesize new elements. Finally, where the right side is shorter than the left side, then `left` is used to synthesize new elements.

## Implementation

The implementation will be similar to `Chunk#zipWith`.

## Examples

Tests should be written to verify functionality and placed into `ChunkSpec.scala`.

```scala
Chunk(1, 2, 3).zipAllWith(Chunk(3, 2, 1))(_ => 0, _ => 0)(_ + _) == Chunk(4, 4, 4)

Chunk(1, 2, 3).zipAllWith(Chunk(3, 2))(_ => 0, _ => 0)(_ + _) == Chunk(4, 4, 3)

Chunk(1, 2).zipAllWith(Chunk(3, 2, 1))(_ => 0, _ => 0)(_ + _) == Chunk(4, 4, 3)
```

**Comment by Vilkina** [comment]
I'll take it

**Comment by ioleo** [comment]
@Vilkina are you still working on this? I'd like to take a stab on this.

**Comment by Vilkina** [comment]
@ioleo, I'm stuck so you can take this one.

**Comment by Vilkina** [comment]
@ioleo, I'm working on it :)

**Comment by kaishh** [comment]
@jdegoes 

The last two tests show incorrect results, should both be `Chunk(4, 4, 0)` due to using left/right element.

```scala
Chunk(1, 2, 3).zipAllWith(Chunk(3, 2))(_ => 0, _ => 0)(_ + _) == Chunk(4, 4, 3)
Chunk(1, 2).zipAllWith(Chunk(3, 2, 1))(_ => 0, _ => 0)(_ + _) == Chunk(4, 4, 3)
```

Misled by monomorphism :D 

```scala
Chunk(1, 2, 3).zipAllWith(Chunk(3, 2, 1))(_ => "L", _ => "R")(_.toString + _.toString) == Chunk("13", "22", "31")

Chunk(1, 2, 3).zipAllWith(Chunk(3, 2))(_ => "L", _ => "R")(_.toString + _.toString) == Chunk("13", "22", "R")

Chunk(1, 2).zipAllWith(Chunk(3, 2, 1))(_ => "L", _ => "R")(_ + _) == Chunk("13", "22", "L")
```

**Comment by ioleo** [comment]
@Vilkina good to hear :+1:

**Comment by Vilkina** [comment]
@ioleo I don't know. Tests are false. So you can take it.

**Comment by DanGrigore** [comment]
I will give it a try also.

**Comment by Vilkina** [comment]
@DanGrigore I'll try one more time ...

**Comment by ghostdogpr** [comment]
Closed by #1612
